File: Python3/46.Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        self.permute_helper(nums, [], ans)
        return ans

    def permute_helper(self, nums, prefix, ans):
        if len(nums) == 0:
            ans.append(prefix.copy())
            return

        for i in range(len(nums)):
            prefix.append(nums[i])
            self.permute_helper(nums[:i] + nums[i + 1:], prefix, ans)
            prefix.pop()

File: Java/46.Permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {

        List<List<Integer>> ans = new ArrayList<>();
        permuteHelper(nums, new ArrayList<>(), ans);
        return ans;

    }

    private void permuteHelper(int[] nums, List<Integer> prefix, List<List<Integer>> ans) {
        if (prefix.size() == nums.length) {
            ans.add(new ArrayList<>(prefix));
            return;
        }

        for (int i = 0; i < nums.length; ++i) {
            if (prefix.contains(nums[i])) {
                continue;
            }
            prefix.add(nums[i]);
            permuteHelper(nums, prefix, ans);
            prefix.remove(prefix.size() - 1);
        }
    }
}